libs/wingc/src/type_check/inference_visitor.rs

```diff
@@ -187,3 +187,32 @@ impl<'a> crate::visit_types::VisitType<'_> for InferenceVisitor<'a> {
 		}
 	}
 }
+
+#[derive(Default)]
+/// Deeply visits a type and finds any inferences.
+pub struct InferenceCounterVisitor {
```

Comment:
ℹ️ I created this during some other experiments while researching the issue. It's not functionally different than the previous visitor for the use-case of looking for inferences. It's just more efficient and has no `mut`s anywhere.